### PR TITLE
Unignore testdata test and make it hermetic

### PR DESCRIPTION
**Pull request by robertbastian**
Based on #4294

### CODE REVIEW

File: provider/datagen/src/tests/make_testdata.rs [hunk 1]
```diff
@@ -18,33 +16,50 @@ use std::sync::Mutex;
...
-    let json_out = Box::new(
-        FilesystemExporter::try_new(Box::new(Json::pretty()), {
-            let mut options = ExporterOptions::default();
-            options.root = "tests/data/json".into();
-            options.overwrite = OverwriteOption::RemoveAndReplace;
-            options
+#[cfg(feature = "use_wasm")]
+fn make_testdata() {
+    // Only produce output if the variable is set. Test is hermetic otherwise.
+    let exporter: Box<dyn DataExporter> = if std::option_env!("TESTDATA_OUT").is_none() {
```
Comment (sffc) [inline provider/datagen/src/tests/make_testdata.rs:22]: Nit: `TESTDATA_OUT` looks like it takes a value of where you write the testdata, but it is only a boolean. How about `ICU4X_WRITE_TESTDATA`

Comment (robertbastian) [reply provider/datagen/src/tests/make_testdata.rs:22]: Sure.

I'm confused that you're leaving a nit comment but not giving approval. Either your comment is blocking, or it's a nit.

Comment (sffc) [reply provider/datagen/src/tests/make_testdata.rs:22]: Ok

**Review by sffc — CHANGES_REQUESTED** [review]

File: provider/datagen/src/tests/make_testdata.rs (hunk 1 above)
Comment (robertbastian) [reply provider/datagen/src/tests/make_testdata.rs:22]: so it's not a nit?

Comment (sffc) [reply provider/datagen/src/tests/make_testdata.rs:22]: I changed my review from non-blocking to blocking

**Review by Manishearth — COMMENTED** [review]
r=me with the environment variable name change

**Comment by robertbastian** [comment]
Mac and Linux are timing out and I have no idea why

File: provider/datagen/src/tests/make_testdata.rs [hunk 2]
```diff
@@ -137,15 +149,15 @@ impl DataExporter for PostcardTestingExporter {
 
         if deallocated != allocated {
             if !EXPECTED_VIOLATIONS.contains(&key) {
-                log::warn!("Zerocopy violation {key} {locale}: {allocated}B allocated, {deallocated}B deallocated");
+                eprintln!("Zerocopy violation {key} {locale}: {allocated}B allocated, {deallocated}B deallocated");
```
Comment (sffc) [inline provider/datagen/src/tests/make_testdata.rs:152]: Question: why did you change this from log::warn to eprintln? log::warn is better because it tells you the filename and line number and you can configure where the log messages go.

**Review by sffc — APPROVED** [review]

**Review by Manishearth — DISMISSED** [review]
this is a nice approach

(no strong opinion on the log stuff)

File: provider/datagen/src/tests/make_testdata.rs (hunk 2 above)
Comment (sffc) [reply provider/datagen/src/tests/make_testdata.rs:152]: Ok, makes sense if we're in a test

**Comment by robertbastian** [comment]
Looks like this was actually a deadlock in wasmer, and #4621 fixed it :)

**Review by sffc — APPROVED** [review]
